README.md
# react-vimeouploadr

[![Travis][travis-badge]][travis]
[![Codecov Status][codecov-badge]][codecov]<!-- [![Github Tag][githubtag-badge]][githubtag] -->
[![Dependency Status][dependency-badge]][dependency]
[![devDependency Status][devdependency-badge]][devdependency]
[![peerDependency Status][peerdependency-badge]][peerdependency]
[![prettier][prettier-badge]][prettier]
[![license][license-badge]][license]

React component for upload video to your vimeo account

## Demo

![demo](http://g.recordit.co/lG2JE1ZXiY.gif)

## Usage

### Steps

1.  Create vimeo app and get the access token (you may need to wait for couple days to get the upload permission)
    [vimeo app register](https://developer.vimeo.com/apps)

    [vimeo upload api doc](https://developer.vimeo.com/api/upload/videos)

2.  You have to prepare a server for requesting vimeo api, you can do this in client side (Well, technically, you can do it on client-side but it will expose your access token to the world and that's not we want to see). Here is the server side sample code:
    [Flask server sample code](https://gist.github.com/ArvinH/42ad574cd8caf6ccd55afa7e2ad9c43a)

3.  Mount component and reducer to your app (Check the `app.js` and `store.js` in `demo/src` foldr)

    ### component

    ```js
    import VimeoUploadr from 'react-vimeouploadr/lib';
    // pass your server host for create video and get video link api (see the servier side code sample)
    <VimeoUploadr
      createVideoLink="http://localhost:5000/video/vimeo"
      getVideoLink="http://localhost:5000/video/vimeo"
    />;
    ```

    ### reducer

    ```js
    import { vimeoUpload } from 'react-vimeouploadr/lib';
    // add reducer to your store
    export default createStore(
      combineReducers({
        vimeoUpload
      }),
      initialState,
      composedEnhancers
    );
    ```

    ### app

    ```js
    // get props to state in component
    const mapStateToProps = state => ({
      vimeoLink: state.vimeoUpload.vimeoLink,
      uploadStatus: state.vimeoUpload.uploadStatus
    });

    DemoApp = connect(mapStateToProps, null)(DemoApp);
    ```

## To-Do

* [x] Pass api host from config
* [x] Export final vimeo video link instead of render player directly
* [x] Doc for api on server side
* [ ] Integrate with Storybook
* [ ] Fix test

[travis-badge]: https://img.shields.io/travis/ArvinH/react-vimeouploadr/master.svg?style=flat-square
[travis]: https://travis-ci.org/ArvinH/react-vimeouploadr
[codecov-badge]: https://img.shields.io/codecov/c/github/ArvinH/react-vimeouploadr.svg?style=flat-square
[codecov]: https://codecov.io/github/ArvinH/react-vimeouploadr?branch=master
[npm-badge]: https://img.shields.io/npm/v/react-vimeouploadr.svg?style=flat-square
[npm]: https://www.npmjs.com/package/react-vimeouploadr
[npm-downloads]: https://img.shields.io/npm/dt/react-vimeouploadr.svg?style=flat-square
[dependency-badge]: https://david-dm.org/ArvinH/react-vimeouploadr.svg?style=flat-square
[dependency]: https://david-dm.org/ArvinH/react-vimeouploadr
[devdependency-badge]: https://david-dm.org/ArvinH/react-vimeouploadr/dev-status.svg?style=flat-square
[devdependency]: https://david-dm.org/ArvinH/react-vimeouploadr#info=devDependencies
[peerdependency-badge]: https://david-dm.org/ArvinH/react-vimeouploadr/peer-status.svg?style=flat-square
[peerdependency]: https://david-dm.org/ArvinH/react-vimeouploadr#info=peerDependencies
[githubtag-badge]: https://img.shields.io/github/tag/ArvinH/react-vimeouploadr.svg?style=flat-square
[githubtag]: ./CHANGELOG.md
[license-badge]: https://img.shields.io/github/license/ArvinH/react-vimeouploadr.svg?style=flat-square
[license]: http://arvinh.mit-license.org/
[prettier-badge]: https://img.shields.io/badge/styled_with-prettier-ff69b4.svg?style=flat-square
[prettier]: https://github.com/prettier/prettier


CHANGELOG.md
# Change Log

All notable changes to the "react-vimeouploadr" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

* Initial release


tests/index-test.js
import expect from 'expect';
import React from 'react';
import Enzyme from 'enzyme';
import { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

describe('Component', () => {

  beforeEach(() => {
  })

  afterEach(() => {
  })

  it('displays a welcome message', () => {
    expect(true).toBe(true);
  })
})


src/reducers/vimeoUpload.js
import axios from 'axios';
import get from 'lodash/get';
// action types
export const POST_VIMEO_START = 'vimeo/POST_VIMEO_START';
export const POST_VIMEO_SUCCESS = 'vimeo/POST_VIMEO_SUCCESS';
export const POST_VIMEO_FAILED = 'vimeo/POST_VIMEO_FAILED';
export const POST_VIMEO_PROGRESS = 'vimeo/POST_VIMEO_PROGRESS';

// initial state
const initialState = {
  vimeoLink: undefined,
  uploading: false,
  uploadStatus: ''
};

// reducer
export default (state = initialState, action) => {
  switch (action.type) {
    case POST_VIMEO_START:
      return {
        ...state,
        ...action.payload
      };

    case POST_VIMEO_SUCCESS:
      return {
        ...state,
        ...action.payload
      };

    case POST_VIMEO_FAILED:
      return {
        ...state,
        ...action.payload
      };

    case POST_VIMEO_PROGRESS:
      return {
        ...state,
        ...action.payload
      };

    default:
      return state;
  }
};

// actions
export const uploadVimeo = ({ createVideoLink, getVideoLink, ...postBody }) => {
  return async dispatch => {
    const success = async (videoUri, videoData, size, uploadLink, result) => {
      const apiResult = result.data;
      // check if upload success
      const checkUploadResult = await axios.head(uploadLink, {
        headers: {
          'Tus-Resumable': '1.0.0'
        }
      });

      const uploadLength = get(checkUploadResult, ['headers', 'upload-length']);
      const uploadOffset = get(checkUploadResult, ['headers', 'upload-offset']);
      if (uploadLength === uploadOffset) {
        // complete upload
        const getVideoExternalLInk = await axios.get(
          `${getVideoLink}${videoUri}`
        );
        dispatch({
          type: POST_VIMEO_SUCCESS,
          payload: {
            vimeoLink: getVideoExternalLInk.data.data,
            uploading: false,
            uploadStatus: 'success'
          }
        });
        return apiResult;
      } else {
        // upload failed, resume upload?
        console.log('should resume upload');
        return patchVimeoFunc(
          videoUri,
          uploadLink,
          videoData,
          size,
          uploadOffset
        );
      }
    };
    const failed = () => {
      dispatch({
        type: POST_VIMEO_FAILED,
        payload: {
          uploading: false,
          uploadStatus: 'failed'
        }
      });
      return null;
    };

    const patchVimeoFunc = async (
      videoUri,
      uploadLink,
      videoData,
      size,
      uploadOffset
    ) => {
      const result = await axios.patch(uploadLink, videoData, {
        headers: {
          'Tus-Resumable': '1.0.0',
          'Upload-Offset': uploadOffset || '0',
          'Content-Type': 'application/offset+octet-stream'
        },
        onUploadProgress: function(progressEvent) {
          const total = progressEvent.total || size;
          const loaded = progressEvent.loaded || 0;
          const progress = loaded / total * 100;
          dispatch({
            type: POST_VIMEO_PROGRESS,
            payload: {
              uploading: true,
              progress: progress >= 100 ? 90 : progress
            }
          });
        }
      });
      return success(videoUri, videoData, size, uploadLink, result);
    };

    try {
      dispatch({
        type: POST_VIMEO_START,
        payload: {
          uploading: true
        }
      });
      const { videoData, size } = postBody;

      // get upload link
      const createVideoAPIResult = await axios.post(createVideoLink, {
        upload: {
          approach: 'tus',
          size: size
        }
      });
      const createVideoResult = createVideoAPIResult.data;
      if (createVideoResult.status !== 'success') {
        return failed();
      }
      const uploadLink = createVideoResult.uploadLink;
      const videoUri = createVideoResult.videoUri;

      // start to upload
      return patchVimeoFunc(videoUri, uploadLink, videoData, size);
    } catch (err) {
      console.log(`post vimeo err: ${err}`);
      return failed();
    }
  };
};


src/components/VimeoUploadr.js
import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import styled from 'styled-components';
import Dropzone from 'react-dropzone';

import { uploadVimeo } from '../reducers/vimeoUpload';

import 'loaders.css/loaders.min.css';

const VideoUploaderWrapper = styled.div`
  width: 100%;
  padding: 0.5em 2em;
  text-align: center;
  background-color: #3a3a3a;
  color: rgba(255, 255, 255, 0.75);
  box-sizing: border-box;
  position: relative;
`;

const DropzoneStyle = {
  width: '200px',
  height: '200px',
  borderWidth: '2px',
  borderColor: 'rgb(102, 102, 102)',
  borderStyle: 'dashed',
  borderRadius: '5px',
  margin: '0 auto',
  position: 'relative',
  overflow: 'hidden'
};

const UplaodingMask = styled.div.attrs({
  className: 'loader-inner ball-pulse'
})`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  padding-top: 50%;
  background-color: rgba(0, 0, 0, 0.6);
`;

const UplaodFailedMessage = styled.div`
  font-color: red;
`;

const ProgressBarWrapper = styled.div`
  width: 200px;
  margin 10px auto;
`;

const ProgressBar = styled.div`
  width: ${props => {
    if (props.uploadStatus === 'success') {
      return 100;
    } else {
      return props.progress || 3;
    }
  }}%;
  height: 15px;
  border-radius: 10px;
  background-color: green;
  transition: all 0.5s;
`;

class VideoUploader extends Component {
  constructor(props) {
    super(props);
    this.state = {
      accepted: [],
      rejected: []
    };
  }

  onDrop = (accepted, rejected) => {
    const { createVideoLink, getVideoLink } = this.props;
    this.setState({
      accepted
    });
    const file = accepted.find(f => f);
    this.props.uploadVimeo({
      createVideoLink,
      getVideoLink,
      videoData: file,
      size: file.size
    });
  };

  render() {
    const {
      VideoUploaderWrapperStyle,
      CustomDropzoneStyle,
      uploading,
      uploadStatus,
      progress
    } = this.props;
    const uploadSuccess = uploadStatus === 'success';
    const uploadFailed = uploadStatus === 'failed';
    return (
      <VideoUploaderWrapper
        className="VideoUploaderWrapper"
        style={VideoUploaderWrapperStyle}
      >
        {uploadFailed && (
          <UplaodFailedMessage>上傳失敗, 請換張圖試試！</UplaodFailedMessage>
        )}
        <Dropzone
          style={{
            ...DropzoneStyle,
            ...CustomDropzoneStyle
          }}
          multiple={false}
          accept="video/mp4"
          onDrop={this.onDrop}
          disabled={uploading}
        >
          <p>
            Try dropping some files here, or click to select files to upload.
          </p>
          {uploading && (
            <UplaodingMask>
              <div />
              <div />
              <div />
            </UplaodingMask>
          )}
        </Dropzone>
        {(uploading || uploadSuccess) && (
          <ProgressBarWrapper>
            <ProgressBar progress={progress} uploadStatus={uploadStatus} />
          </ProgressBarWrapper>
        )}
      </VideoUploaderWrapper>
    );
  }
}

const mapStateToProps = state => ({
  uploading: state.vimeoUpload.uploading,
  uploadStatus: state.vimeoUpload.uploadStatus,
  progress: state.vimeoUpload.progress
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      uploadVimeo
    },
    dispatch
  );
VideoUploader = connect(mapStateToProps, mapDispatchToProps)(VideoUploader);

export default VideoUploader;


demo/src/app.js
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { render } from 'react-dom';
import styled from 'styled-components';
import VimeoUploadr from '../../src';


const VideoComponent = styled.video`
  display: inline-block;
  width: 50%;
  max-height: 600px;
  box-sizing: border-box;
  vertical-align: middle;
`;

class DemoApp extends Component {
  render() {
    const {
      vimeoLink,
      uploadStatus
    } = this.props;
    const uploadSuccess = uploadStatus === 'success';
    return (
      <div>
        <VimeoUploadr
          createVideoLink="http://localhost:5000/video/vimeo"
          getVideoLink="http://localhost:5000/video/vimeo"
        />
        {uploadSuccess && <VideoComponent src={vimeoLink} controls />}
      </div>
    );
  }
};

const mapStateToProps = state => ({
  vimeoLink: state.vimeoUpload.vimeoLink,
  uploadStatus: state.vimeoUpload.uploadStatus,
});

DemoApp = connect(
  mapStateToProps,
  null
)(DemoApp);

export default DemoApp;